x-pack/plugins/enterprise_search/public/applications/app_search/components/role_mappings/role_mapping.tsx

```diff
@@ -0,0 +1,292 @@
+  DeleteMappingCallout,
+  RoleSelector,
+} from '../../../shared/role_mapping';
+import { ROLE_MAPPINGS_TITLE } from '../../../shared/role_mapping/constants';
+import { AppLogic } from '../../app_logic';
+import { RoleTypes } from '../../types';
+import { roleHasScopedEngines } from '../../utils/role/has_scoped_engines';
+import { Engine } from '../engine/types';
+
+import {
+  DEV_ROLE_TYPE_DESCRIPTION,
+  EDITOR_ROLE_TYPE_DESCRIPTION,
+  ANALYST_ROLE_TYPE_DESCRIPTION,
+  OWNER_ROLE_TYPE_DESCRIPTION,
+  ADMIN_ROLE_TYPE_DESCRIPTION,
+  ADVANCED_ROLE_SELECTORS_TITLE,
+  ROLE_TITLE,
+  FULL_ENGINE_ACCESS_TITLE,
+  FULL_ENGINE_ACCESS_DESCRIPTION,
+  LIMITED_ENGINE_ACCESS_TITLE,
+  LIMITED_ENGINE_ACCESS_DESCRIPTION,
+  ENGINE_ACCESS_TITLE,
+} from './constants';
+import { RoleMappingsLogic } from './role_mappings_logic';
+
+interface RoleMappingProps {
+  isNew?: boolean;
+}
+
+interface AdvanceRoleType {
+  type: RoleTypes;
+  description: string;
+}
+
+const advancedRoleTypes = [
+  {
+    type: 'dev',
+    description: DEV_ROLE_TYPE_DESCRIPTION,
+  },
+  {
+    type: 'editor',
+    description: EDITOR_ROLE_TYPE_DESCRIPTION,
+  },
+  {
+    type: 'analyst',
+    description: ANALYST_ROLE_TYPE_DESCRIPTION,
+  },
+] as AdvanceRoleType[];
+
+const standardRoleTypes = [
+  {
+    type: 'owner',
+    description: OWNER_ROLE_TYPE_DESCRIPTION,
+  },
+  {
+    type: 'admin',
+    description: ADMIN_ROLE_TYPE_DESCRIPTION,
+  },
+] as AdvanceRoleType[];
+
+export const RoleMapping: React.FC<RoleMappingProps> = ({ isNew }) => {
+  const { roleId } = useParams() as { roleId: string };
+  const { myRole } = useValues(AppLogic);
+
+  const {
+    handleAccessAllEnginesChange,
+    handleAttributeSelectorChange,
+    handleAttributeValueChange,
+    handleAuthProviderChange,
+    handleDeleteMapping,
+    handleEngineSelectionChange,
+    handleRoleChange,
+    handleSaveMapping,
+    initializeRoleMapping,
+    resetState,
+  } = useActions(RoleMappingsLogic);
+
+  const {
+    accessAllEngines,
+    attributeName,
+    attributeValue,
+    attributes,
+    availableAuthProviders,
+    availableEngines,
+    dataLoading,
+    elasticsearchRoles,
+    hasAdvancedRoles,
+    multipleAuthProvidersConfig,
+    roleMapping,
+    roleType,
+    selectedEngines,
+    selectedAuthProviders,
+  } = useValues(RoleMappingsLogic);
+
+  useEffect(() => {
+    initializeRoleMapping(roleId);
+    return resetState;
+  }, []);
+
+  if (dataLoading) return <Loading />;
+
+  const SAVE_ROLE_MAPPING_LABEL = i18n.translate(
+    'xpack.enterpriseSearch.appSearch.roleMapping.saveRoleMappingButtonMessage',
+    {
+      defaultMessage: '{operation} role mapping',
+      values: { operation: isNew ? 'Save' : 'Update' },
+    }
+  );
+
+  const ACTIVE_BREADCRUMB_TEXT = i18n.translate(
+    'xpack.enterpriseSearch.appSearch.roleMapping.activeBreadcrumbText',
+    {
+      defaultMessage: '{operation} role mapping',
+      values: { operation: isNew ? 'Add' : 'Manage' },
+    }
+  );
```

Comment:
We can't have conditionals in our i18n values unfortunately - the words Save/Update/Add/Manage won't actually get properly translated. I strongly recommend pulling these strings out to a `constants` file and doing something like this instead:

```js
// constants.ts

const SAVE_ROLE_MAPPING = i18n.translate(
  'xpack.enterpriseSearch.appSearch.roleMapping.saveRoleMappingButtonLabel',
  { defaultMessage: 'Save role mapping' }
);
const UPDATE_ROLE_MAPPING = i18n.translate(
  'xpack.enterpriseSearch.appSearch.roleMapping.updateRoleMappingButtonLabel',
  { defaultMessage: 'Update role mapping' }
);

const ADD_ROLE_MAPPING_TITLE = i18n.translate(
  'xpack.enterpriseSearch.appSearch.roleMapping.newRoleMappingTitle',
  { defaultMessage: 'Add role mapping' }
);
const MANAGE_ROLE_MAPPING_TITLE = i18n.translate(
  'xpack.enterpriseSearch.appSearch.roleMapping.manageRoleMappingTitle',
  { defaultMessage: 'Manage role mapping' }
);
```

And then here, you can do something like:

```suggestion
const SAVE_ROLE_MAPPING_LABEL = isNew ? SAVE_ROLE_MAPPING : UPDATE_ROLE_MAPPING;
const TITLE = isNew ? ADD_ROLE_MAPPING_TITLE : MANAGE_ROLE_MAPPING_TITLE;
```

Reply:
https://github.com/elastic/kibana/pull/94763/commits/d7e029b9a54938716cb4a9ed921fb1dbe0ac86a9